dockerfile_parse/util.py

```diff
@@ -188,3 +188,105 @@ def remove_nonescaped_quotes(string):
+
+
+def extract_labels_or_envs(env_replace, envs, instruction_value):
+    shlex_splits_raw = shlex_split(instruction_value,
+                                   env_replace=env_replace, envs=envs)
+
+    key_val_list = []
+    if '=' not in shlex_splits_raw[0]:  # LABEL/ENV name value
+        # split it to first (name) and the rest (value)
+        key_val = instruction_value.split(None, 1)
+        key = strip_quotes(key_val[0])
+        try:
+            val = key_val[1]
+        except IndexError:
+            val = ''
+
+        if env_replace:
+            val = EnvSubst(val, envs).substitute()
+        val = remove_nonescaped_quotes(val)
+
+        key_val_list.append((key, val))
+
+    else:  # LABEL/ENV "name"="value"
+        shlex_splits = []
+        append_to_last = False
+        for part in shlex_splits_raw:
+            if append_to_last:
+                shlex_splits[-1] += part
+                append_to_last = False
+                continue
+            if part != '=':
+                shlex_splits.append(part)
+            else:
+                shlex_splits[-1] += "="
+                append_to_last = True
+
+        for part in shlex_splits:
+            key, val = part.split('=')
+            key_val_list.append((key, val))
+
+    return key_val_list
+
+
+def get_key_val_dictionary(instruction_value, env_replace=False, envs=None):
+    envs = envs or []
+    return dict(extract_labels_or_envs(instruction_value=instruction_value,
+                                       env_replace=env_replace,
+                                       envs=envs))
+
+
+class Context(object):
+    def __init__(self, envs=None, labels=None, line_envs=None, line_labels=None):
+        """
+        Class representing current state of environment variables and labels.
+
+        :param envs: dict with variables valid for this line
+            (all variables defined to this line)
+        :param labels: dict with labels valid for this line
+            (all labels defined to this line)
+        :param line_envs: dict with variables defined on this line
+        :param line_labels: dict with labels defined on this line
+        """
+        self.envs = envs or {}
+        self.labels = labels or {}
+        self.line_envs = line_envs or {}
+        self.line_labels = line_labels or {}
+
+    def update_line_value(self, context_type, value):
+        """
+        Update value defined on this line.
+
+        :param context_type: "ENV" or "LABEL"
+        :param value: new value for this line
+        """
+        if context_type.upper() == "ENV":
+            self.line_envs = value
+        elif context_type.upper() == "LABEL":
+            self.line_labels = value
+
+    def get_line_value(self, context_type):
+        """
+        Get the values defined on this line.
+
+        :param context_type: "ENV" or "LABEL"
+        :return: values of given type defined on this line
+        """
```

Comment:
I agree that the getters seem overkill. Using `context.envs` versus `context.get_values('ENV')` is a much better interface IMO.

I do see the benefit of `set_line_value`, but am wondering if this would be better handled by passing the instruction to `Context.__init__` for better encapsulation. Either way, I'm also fine with the existing setter approach.